### PR TITLE
Virtual drive

### PR DESCRIPTION
**Pull request by camilasan**
**Summary**
The Virtual Drive is a feature expansion of the Nextcloud Desktop Client. It offers a virtual disk drive as replacement for the locally synced folders currently used by the client. The Virtual Drive is available to the operating system while the client is running and shows all files the user has in their Nextcloud account. Files are retrieved on-demand, when the user tries to open a file. Data is cached based on (initially) a date-based algorithm, removing files after a set time of no use. In the future alternative algorithms can be offered, like evicting the least recently used file based on a user defined size constraint.

### Features
* Virtual disk drive available while client is running
* Caches data locally for performance and off-line use
* Syncs before files are accessed when online
* Automatically retires cached files after a set time of going unused
* Full compatibility with any application
* Users can manually select files as available offline
* Right click menu options include the options available in the current client
* Available optionally in Windows and on Mac OS

### User experience
* Cached files should open without a noticeable delay (under 1 second in typical circumstances)
* The drive should look and act like a locally available file storage

### Design (WIP)
* The client creates a local, hidden folder used to cache the files - it is in the folder where we put the database and sync log files.
* FUSE then mounts the local, hidden folder and shows it as a virtual drive to the user
* The client connects to the server and gets all folders and files structure into the hidden folder with zero byte files that the user can open with his applications
* The zero byte files are marked in the database with a virtualfile flag 1 
* When the user opens/edit/move/remove a file, this change will be immediatly synced 
* then virtualfile flag is set to 0

## Roadmap

Note that this is specifically about the Virtual Drive feature, not generic Desktop Client capabilities.

* Linux support
* Other cache cleaning methods like based on size limitations
* Automatically keeping favorite and recommended files offline
* Replace the current algorithm for change detection to use a recorded list of changes to speed up syncing

## TO DO
- [ ] the metadata db table stores if the file is a virtualfile (1) or not (0) - virtualfiles are 0 byte files created by FUSE so the user can see in his mounted drive which files are available in the server
~~ctx->priority.file contains the files the user touched (opened or whatever) that needs to be synced independent of being offline files or not. ctx->priority.file should also determine what is priority to be synced by the sync engine~~
- [ ] a file that is offline and is a virtualfile or a file that has changed (opened by the user) and is offline should be priority in the syncing 
- [ ] update the vfs_mac and vfs_maccontroller with the last changes made that connect FUSE with the sync engine.
- [ ] Linux....
- [x] more rebase
- [x] remove the option to enble/disable the virtual drive from gui - have it disabled by default, only available to change via the config file
- [x] rebase to master

### CODE REVIEW

**Comment by er-vin** [comment]
Rebased! Woohoo! This is awesome, now the long review period shall begin. ;-)

File: src/gui/socketapi.cpp [hunk 1]
```diff
@@ -366,8 +396,20 @@ void SocketApi::broadcastMessage(const QString &msg, bool doWait)
...
 }
 
-void SocketApi::processShareRequest(const QString &localFile, SocketListener *listener, ShareDialogStartPage startPage)
+void SocketApi::processShareRequest(const QString &localFileC, SocketListener *listener, ShareDialogStartPage startPage)
 {
+    QString localFile = localFileC;
+
+#if defined(Q_OS_WIN)
+    OCC::ConfigFile Cfg;
+    char letter[2];
```
Comment (er-vin) [inline src/gui/socketapi.cpp:405]: It is wrapped back into a QString later on, so why not just get the first QChar of that string?

File: src/gui/socketapi.cpp [hunk 2]
```diff
@@ -366,8 +396,20 @@ void SocketApi::broadcastMessage(const QString &msg, bool doWait)
...
+    letter[0] = localFile.toStdString().c_str()[0];
+    letter[1] = 0;
+
+    if (!QString(letter).compare(Cfg.defaultFileStreamLetterDrive().toUpper()))
```
Comment (er-vin) [inline src/gui/socketapi.cpp:409]: If we assume it's always a single letter (as the code above implies), defaultFileStreamLetterDrive could return a QChar then I guess.

But is it always a single letter?

Also that whole logic seems to be present plenty of times in that file, probably needs to be factored out.

File: src/gui/socketapi.cpp [hunk 3]
```diff
@@ -411,9 +453,23 @@ void SocketApi::broadcastStatusPushMessage(const QString &systemPath, SyncFileSt
 {
     QString msg = buildMessage(QLatin1String("STATUS"), systemPath, fileStatus.toSocketAPIString());
     Q_ASSERT(!systemPath.endsWith('/'));
+#if defined(Q_OS_WIN)
+	ConfigFile Cfg;
```
Comment (er-vin) [inline src/gui/socketapi.cpp:457]: Please fix indentation

File: src/gui/socketapi.cpp [hunk 4]
```diff
@@ -411,9 +453,23 @@ void SocketApi::broadcastStatusPushMessage(const QString &systemPath, SyncFileSt
...
     foreach (auto &listener, _listeners) {
-        listener.sendMessageIfDirectoryMonitored(msg, directoryHash);
+#if defined(Q_OS_WIN)
+        QString relative_prefix = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) + "/cachedFiles/";
+        QString systemPath2 = systemPath;
+        systemPath2.replace(0, relative_prefix.length(), Cfg.defaultFileStreamLetterDrive().toUpper().append(":"));
+        QString msg2 = buildMessage(QLatin1String("STATUS"), systemPath2, fileStatus.toSocketAPIString());
+		listener.sendMessageIfDirectoryMonitored(msg2, directoryHash);
+#elif defined(Q_OS_MAC)
+		listener.sendMessageIfDirectoryMonitored(msg, directoryHash);
```
Comment (er-vin) [inline src/gui/socketapi.cpp:471]: Please fix indentation

File: src/gui/socketapi.cpp [hunk 5]
```diff
@@ -667,50 +858,133 @@ void SocketApi::fetchPrivateLinkUrlHelper(const QString &localFile, const std::f
...
-        { "CONTEXT_MENU_TITLE", Theme::instance()->appNameGUI() },
+        { "CONTEXT_MENU_TITLE", tr("Share via ") + Theme::instance()->appNameGUI()},
         { "COPY_PRIVATE_LINK_MENU_TITLE", tr("Copy private link to clipboard") },
-        { "EMAIL_PRIVATE_LINK_MENU_TITLE", tr("Send private link by email …") },
-        { "CONTEXT_MENU_ICON", APPLICATION_ICON_NAME},
+        { "EMAIL_PRIVATE_LINK_MENU_TITLE", tr("Send private link by email...") },
     } };
     listener->sendMessage(QString("GET_STRINGS:BEGIN"));
-    for (const auto& key_value : strings) {
+    for (auto key_value : strings) {
```
Comment (er-vin) [inline src/gui/socketapi.cpp:987]: Should still be a const ref IMO.

File: src/gui/socketapi.cpp [hunk 6]
```diff
@@ -731,10 +1005,14 @@ void SocketApi::sendSharingContextMenuOptions(const FileData &fileData, SocketLi
 
     // If sharing is globally disabled, do not show any sharing entries.
     // If there is no permission to share for this file, add a disabled entry saying so
+
+    listener->sendMessage(QLatin1String("MENU_ITEM:OFFLINE_DOWNLOAD_MODE") + flagString + tr("0ffline"));
```
Comment (er-vin) [inline src/gui/socketapi.cpp:1009]: Offline not 0ffline I guess. ;-)

File: src/gui/socketapi.cpp [hunk 7]
```diff
@@ -755,11 +1033,23 @@ void SocketApi::sendSharingContextMenuOptions(const FileData &fileData, SocketLi
 
     // Disabled: only providing email option for private links would look odd,
     // and the copy option is more general.
-    //listener->sendMessage(QLatin1String("MENU_ITEM:EMAIL_PRIVATE_LINK") + flagString + tr("Send private link by email …"));
+    //listener->sendMessage(QLatin1String("MENU_ITEM:EMAIL_PRIVATE_LINK") + flagString + tr("Send private link by email..."));
```
Comment (er-vin) [inline src/gui/socketapi.cpp:1036]: Unneeded change

File: src/gui/socketapi.cpp [hunk 8]
```diff
@@ -792,25 +1082,34 @@ SyncJournalFileRecord SocketApi::FileData::journalRecord() const
...
+    letter[1] = 0;
+
+    if (!QString(letter).compare(Cfg.defaultFileStreamLetterDrive().toUpper()))
+        argument.replace(0, 3, QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) + "/cachedFiles/");
+#endif
+
     listener->sendMessage(QString("GET_MENU_ITEMS:BEGIN"));
-    bool hasSeveralFiles = argument.contains(QLatin1Char('\x1e')); // Record Separator
+    auto flagString = QLatin1String(":d:");
+	listener->sendMessage(QLatin1String("MENU_ITEM:OFFLINE_DOWNLOAD_MODE") + flagString + tr("0ff line"));
```
Comment (er-vin) [inline src/gui/socketapi.cpp:1101]: Please fix indentation

File: src/gui/socketapi.h [hunk 9]
```diff
@@ -59,6 +59,7 @@ public slots:
 
 signals:
     void shareCommandReceived(const QString &sharePath, const QString &localPath, ShareDialogStartPage startPage);
+	void shareUserGroupCommandReceived(const QString &sharePath, const QString &localPath, bool resharingAllowed);
```
Comment (er-vin) [inline src/gui/socketapi.h:62]: Please fix indentation

File: src/gui/socketapi.h [hunk 10]
```diff
@@ -96,8 +97,9 @@ private slots:
 
     Q_INVOKABLE void command_VERSION(const QString &argument, SocketListener *listener);
 
+    Q_INVOKABLE void command_SHARE_STATUS(const QString &localFile, SocketListener *listener);
     Q_INVOKABLE void command_SHARE_MENU_TITLE(const QString &argument, SocketListener *listener);
-
+
```
Comment (er-vin) [inline src/gui/socketapi.h:102]: No spaces at end of line please.

File: src/gui/socketapi.h [hunk 11]
```diff
@@ -136,10 +141,14 @@ private slots:
     DirectEditor* getDirectEditorForLocalFile(const QString &localFile);
 
     QString buildRegisterPathMessage(const QString &path);
+	QString buildRegisterFsMessage();
```
Comment (er-vin) [inline src/gui/socketapi.h:144]: Please fix indentation

File: src/gui/socketapi.h [hunk 12]
```diff
@@ -136,10 +141,14 @@ private slots:
...
+    Q_INVOKABLE void command_GET_DOWNLOAD_MODE(const QString& localFile, SocketListener* listener);
+
 
     QSet<QString> _registeredAliases;
     QList<SocketListener> _listeners;
     SocketApiServer _localServer;
 };
 }
-#endif // SOCKETAPI_H
+#endif // SOCKETAPI_H
```
Comment (er-vin) [inline src/gui/socketapi.h:154]: Unneeded change

File: src/libsync/configfile.cpp [hunk 13]
```diff
@@ -899,6 +904,80 @@ void ConfigFile::setAutomaticLogDir(bool enabled)
     settings.setValue(QLatin1String(automaticLogDirC), enabled);
 }
 
+void ConfigFile::setDefaultFileStreamMirrorPath(QString path)
+{
+	QSettings settings(configFile(), QSettings::IniFormat);
```
Comment (er-vin) [inline src/libsync/configfile.cpp:909]: Please fix indentation

File: src/libsync/configfile.h [hunk 14]
```diff
@@ -83,6 +83,18 @@ class OWNCLOUDSYNC_EXPORT ConfigFile
     bool monoIcons() const;
     void setMonoIcons(bool);
 
+// Configuration for virtual FileSystem.
+	void setEnableVirtualFileSystem(bool);
```
Comment (er-vin) [inline src/libsync/configfile.h:87]: Please fix indentation

File: src/libsync/discoveryphase.cpp [hunk 15]
```diff
@@ -734,4 +734,72 @@ void DiscoveryJob::start()
     emit finished(ret);
     deleteLater();
 }
+    void DiscoveryFolderFileList::doGetFolderContent(const QString &subPath)
```
Comment (er-vin) [inline src/libsync/discoveryphase.cpp:737]: Please fix indentation

File: src/libsync/owncloudpropagator.cpp [hunk 16]
```diff
@@ -967,6 +967,7 @@ void PropagateDirectory::slotSubJobsFinished(SyncFileItem::Status status)
                 && _item->_originalFile != _item->_renameTarget) {
                 // Remove the stale entries from the database.
                 propagator()->_journal->deleteFileRecord(_item->_originalFile, true);
+				propagator()->_journal->deleteSyncMode(_item->_originalFile);
```
Comment (er-vin) [inline src/libsync/owncloudpropagator.cpp:970]: Please fix indentation

File: src/libsync/propagatedownload.cpp [hunk 17]
```diff
@@ -901,7 +901,13 @@ void PropagateDownloadFile::downloadFinished()
         // the discovery phase and now.
         const qint64 expectedSize = _item->_previousSize;
         const time_t expectedMtime = _item->_previousModtime;
-        if (!FileSystem::verifyFileUnchanged(fn, expectedSize, expectedMtime)) {
+
```
Comment (er-vin) [inline src/libsync/propagatedownload.cpp:904]: Please fix indentation

File: src/libsync/propagateremotedelete.cpp [hunk 18]
```diff
@@ -156,6 +156,7 @@ void PropagateRemoteDelete::slotDeleteJobFinished()
     }
 
     propagator()->_journal->deleteFileRecord(_item->_originalFile, _item->isDirectory());
+	propagator()->_journal->deleteSyncMode(_item->_originalFile);
```
Comment (er-vin) [inline src/libsync/propagateremotedelete.cpp:159]: Please fix indentation

I also wonder if deleteFileRecord shouldn't do this by itself.

File: src/libsync/propagateremotemove.cpp [hunk 19]
```diff
@@ -149,6 +149,7 @@ void PropagateRemoteMove::finalize()
     // The db is only queried to transfer the content checksum from the old
     // to the new record. It is not a problem to skip it here.
     propagator()->_journal->deleteFileRecord(_item->_originalFile);
+	propagator()->_journal->deleteSyncMode(_item->_originalFile);
```
Comment (er-vin) [inline src/libsync/propagateremotemove.cpp:152]: Please fix indentation

File: src/libsync/propagatorjobs.cpp [hunk 20]
```diff
@@ -85,6 +85,7 @@ bool PropagateLocalRemove::removeRecursively(const QString &path)
             foreach (const auto &it, deleted) {
                 propagator()->_journal->deleteFileRecord(_item->_originalFile + path + QLatin1Char('/') + it.first,
                     it.second);
+				propagator()->_journal->deleteSyncMode(_item->_originalFile + path + QLatin1Char('/') + it.first);
```
Comment (er-vin) [inline src/libsync/propagatorjobs.cpp:88]: Please fix indentation

File: src/libsync/propagatorjobs.cpp [hunk 21]
```diff
@@ -96,6 +97,7 @@ bool PropagateLocalRemove::removeRecursively(const QString &path)
             // This succeeded, so we need to delete it from the database now because the caller won't
             propagator()->_journal->deleteFileRecord(_item->_originalFile + path + QLatin1Char('/') + di.fileName(),
                 isDir);
+			propagator()->_journal->deleteSyncMode(_item->_originalFile + path + QLatin1Char('/') + di.fileName());
```
Comment (er-vin) [inline src/libsync/propagatorjobs.cpp:100]: Please fix indentation

File: src/libsync/propagatorjobs.cpp [hunk 22]
```diff
@@ -149,6 +151,7 @@ void PropagateLocalRemove::start()
     }
     propagator()->reportProgress(*_item, 0);
     propagator()->_journal->deleteFileRecord(_item->_originalFile, _item->isDirectory());
+	propagator()->_journal->deleteSyncMode(_item->_originalFile);
```
Comment (er-vin) [inline src/libsync/propagatorjobs.cpp:154]: Please fix indentation

File: src/libsync/syncengine.cpp [hunk 23]
```diff
@@ -100,6 +100,8 @@ SyncEngine::SyncEngine(AccountPtr account, const QString &localPath,
     _clearTouchedFilesTimer.setInterval(30 * 1000);
     connect(&_clearTouchedFilesTimer, &QTimer::timeout, this, &SyncEngine::slotClearTouchedFiles);
 
+	_csync_ctx->virtualDriveEnabled = ConfigFile().enableVirtualFileSystem();
```
Comment (er-vin) [inline src/libsync/syncengine.cpp:103]: Please fix indentation

File: src/libsync/syncfileitem.cpp [hunk 24]
```diff
@@ -36,6 +36,7 @@ SyncJournalFileRecord SyncFileItem::toSyncJournalFileRecordWithInode(const QStri
     rec._serverHasIgnoredFiles = _serverHasIgnoredFiles;
     rec._checksumHeader = _checksumHeader;
     rec._e2eMangledName = _encryptedFileName.toUtf8();
+	rec._virtualfile = _virtualfile;
```
Comment (er-vin) [inline src/libsync/syncfileitem.cpp:39]: Please fix indentation

File: src/libsync/syncfileitem.h [hunk 25]
```diff
@@ -101,6 +101,15 @@ class SyncFileItem
...
         , _isRestoration(false)
+        , _httpErrorCode(0)
+        , _affectedItems(1)
+        , _instruction(CSYNC_INSTRUCTION_NONE)
+        , _modtime(0)
+        , _size(0)
+        , _inode(0)
+        , _previousSize(0)
+        , _previousModtime(0)
+		, _virtualfile(1)
```
Comment (er-vin) [inline src/libsync/syncfileitem.h:112]: Please fix indentation

File: src/libsync/syncfileitem.h [hunk 26]
```diff
@@ -242,11 +251,13 @@ class SyncFileItem
...
     // The size and modtime of the file getting overwritten (on the disk for downloads, on the server for uploads).
-    quint64 _previousSize = 0;
-    time_t _previousModtime = 0;
+    quint64 _previousSize;
+    time_t _previousModtime;
 
     QString _directDownloadUrl;
     QString _directDownloadCookies;
+	// virtual drive
+	quint64 _virtualfile;
```
Comment (er-vin) [inline src/libsync/syncfileitem.h:260]: Please fix indentation

File: shell_integration/MacOSX/OwnCloudFinderSync/FinderSyncExt/FinderSync.m [hunk 27]
```diff
@@ -137,14 +137,15 @@ - (NSMenu *)menuForMenuKind:(FIMenuKind)whichMenu
 			[actionItem setTag:idx];
 			[actionItem setTarget:self];
 			NSString *flags = [item valueForKey:@"flags"]; // e.g. "d"
-			if ([flags rangeOfString:@"d"].location != NSNotFound) {
-				[actionItem setEnabled:false];
-			}
+			//if ([flags rangeOfString:@"d"].location != NSNotFound) {
```
[the remark this answers is not in the document]
Comment (camilasan) [reply shell_integration/MacOSX/OwnCloudFinderSync/FinderSyncExt/FinderSync.m:140]: I would like to leave commented code for now.

[the remark this answers is not in the document]
Comment (er-vin) [reply shell_integration/MacOSX/OwnCloudFinderSync/FinderSyncExt/FinderSync.m:140]: Fair enough, let's keep it. I wasn't sure if that was intended or an oversight anyway.

File: shell_integration/windows/OCUtil/RemotePathChecker.h [hunk 28]
```diff
@@ -40,8 +40,11 @@ class __declspec(dllexport) RemotePathChecker {
     ~RemotePathChecker();
     std::shared_ptr<const std::vector<std::wstring>> WatchedDirectories() const;
     bool IsMonitoredPath(const wchar_t* filePath, int* state);
+	void setLetterDrive(std::wstring str);
```
[the remark this answers is not in the document]
Comment (camilasan) [reply shell_integration/windows/OCUtil/RemotePathChecker.h]: It was actually not implemented so I removed it.

File: src/common/syncjournaldb.cpp [hunk 29]
```diff
@@ -102,6 +104,14 @@ SyncJournalDb::SyncJournalDb(const QString &dbFilePath, QObject *parent)
     if (_journalMode.isEmpty()) {
         _journalMode = defaultJournalMode(_dbFile);
     }
+
+	ASSERT(!_instance);
+	_instance = this;
+}
+
+SyncJournalDb *SyncJournalDb::instance()
```
[the remark this answers is not in the document]
Comment (camilasan) [reply src/common/syncjournaldb.cpp:112]: Yes, I know.

File: src/common/syncjournaldb.cpp [hunk 30]
```diff
@@ -583,6 +605,48 @@ bool SyncJournalDb::checkConnect()
...
+   /* _setErrorBlacklistQuery.reset(new SqlQuery(_db));
+    if (_setErrorBlacklistQuery->prepare("INSERT OR REPLACE INTO blacklist "
+                                         "(path, lastTryEtag, lastTryModtime, retrycount, errorstring, lastTryTime, ignoreDuration, renameTarget, errorCategory) "
+                                         "VALUES ( ?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)")) {
+        return sqlFail("prepare _setErrorBlacklistQuery", *_setErrorBlacklistQuery);
+    }*/
+
+	// Sync mode
+
+	if (!_getSyncModeDownloadQuery.initOrReset("SELECT downloaded FROM syncmode WHERE path=?1;", _db)) {
```
[the remark this answers is not in the document]
Comment (camilasan) [reply src/common/syncjournaldb.cpp]: I don't think this should be here.

File: src/gui/socketapi.h [hunk 31]
```diff
@@ -48,7 +48,7 @@ class SocketApi : public QObject
     Q_OBJECT
 
 public:
-    explicit SocketApi(QObject *parent = nullptr);
+    explicit SocketApi(QObject *parent = 0);
```
Comment (lexxxel) [inline src/gui/socketapi.h:51]: Isn't a nullptr better then 0?

File: src/gui/vfs_maccontroller.h [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+#include "accountstate.h"
+
+class VfsMac;
+
+class VfsMacController : public QObject
+{
+    Q_OBJECT
+public:
+    explicit VfsMacController(QString rootPath, QString mountPath, OCC::AccountState *accountState, QObject *parent);
+    //~LoopbackController();
```
Comment (lexxxel) [inline src/gui/vfs_maccontroller.h:39]: Wrong Class Name, also who releases and qi_ if there is just a default destructor?

File: src/gui/socketapi.h (hunk 31 above)
Comment (er-vin) [reply src/gui/socketapi.h:51]: Indeed

File: src/gui/vfs_maccontroller.h (hunk 32 above)
Comment (er-vin) [reply src/gui/vfs_maccontroller.h:39]: Indeed for the class name, and the include guards are wrong as well.

qi_ inherits QObject and is properly parented so it'll be deleted as expected in the QObject dtor, no leakage here.

File: src/gui/quotainfo.cpp [hunk 33]
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) by Daniel Molkentin <danimo@owncloud.com>
```
[the remark this answers is not in the document]
Comment (camilasan) [reply src/gui/quotainfo.cpp:2]: File is removed from master.  (it was refactored out)

File: src/gui/socketapi.cpp (hunk 6 above)
Comment (camilasan) [reply src/gui/socketapi.cpp:1008]: ?

Comment (er-vin) [reply src/gui/socketapi.cpp:1008]: Look at the translated string, it has a zero instead of an uppercase o

**Comment by Erisa** [comment]
Are there any (automatic?) builds of this that can be used for testing aside from the old tech preview ones? Attempting to compile it as someone with little experience is proving incredibly difficult.

**Comment by camilasan** [comment]
We have moved the work to https://github.com/nextcloud/desktop/tree/virtual-drive-techpreview

**Comment by simonfranzen** [comment]
Hey, where to find and how to enable the Virtual Drive feature? Isn't it included in the latest releases?

**Comment by FlexW** [comment]
@simonfranzen Have a look here https://nextcloud.com/blog/nextcloud-desktop-client-3-2-with-status-feature-and-virtual-files-available-now/ . Please also look at the first comment from jospoortvliet

**Pull request closed without merging**